### PR TITLE
fix(parser): use Pydantic validation for outputTokens extraction (#810)

### LINKED ISSUE #810: [aw][code health] `parser`: `_safe_int_tokens` can silently discard valid token counts when `outputTokens` is a JSON float
## Root Cause

Both `_first_pass` and `_detect_resume` in `parser.py` read `outputTokens` from `assistant.message` events using raw dict access:

```python
# parser.py – _first_pass (line ~503) and _detect_resume (line ~546)
if (tokens := _safe_int_tokens(ev.data.get("outputTokens"))) is not None:
    total_output_tokens += tokens
```

`_safe_int_tokens` accepts only `isinstance(raw, int) and not isinstance(raw, bool) and raw >= 0`:

```python
def _safe_int_tokens(raw: object) -> int | None:
    """Return *raw* as non-negative int if it is a genuine integer (not bool), else None."""
    if isinstance(raw, int) and not isinstance(raw, bool) and raw >= 0:
        return raw
    return None
```

If `outputTokens` arrives as a JSON float (`1234.0`) — which is a valid JSON number and is parsed by Python's `json` module as a `float` — `_safe_int_tokens` returns `None` and the tokens are **silently discarded**. No warning is emitted.

By contrast, calling `ev.as_assistant_message()` invokes Pydantic validation, which coerces `float → int` (lax mode default for `model_validate_json`) and would return `1234` correctly.

This creates a silent inconsistency between the fast path and the typed-accessor path: the two routes produce different token totals for the same event data. Any future version of the Copilot CLI that emits `outputTokens` as a floating-point literal would produce under-counted token totals with no error signal.

This also violates the coding guideline: **"No `isinstance` checks in business logic"** (`_first_pass` is core business logic, not an I/O boundary coercion).

## Scope

- `parser._safe_int_tokens` — the validator itself
- `parser._first_pass` — calls `_safe_int_tokens(ev.data.get("outputTokens"))`
- `parser._detect_resume` — same pattern

## Fix

**Option A (minimal):** Broaden `_safe_int_tokens` to accept whole-number floats:

```python
def _safe_int_tokens(raw: object) -> int | None:
    if isinstance(raw, bool):
        return None
    if isinstance(raw, int) and raw >= 0:
        return raw
    if isinstance(raw, float) and raw >= 0 and raw == int(raw):
        return int(raw)
    return None
```

**Option B (preferred, guideline-aligned):** Create a lightweight typed helper that extracts `outputTokens` via the Pydantic model and falls back gracefully:

```python
def _extract_output_tokens(ev: SessionEvent) -> int | None:
    try:
        return ev.as_assistant_message().outputTokens or None
    except (ValidationError, ValueError):
        return None
```

Replace both `_safe_int_tokens(ev.data.get("outputTokens"))` call sites with `_extract_output_tokens(ev)`.

Option B aligns with the design intent of the typed `as_*()` accessors and eliminates the `isinstance` check from business logic entirely. The performance cost (one Pydantic validation per `assistant.message` event) should be measured against the baseline before merging.

## Testing

- Add a regression test in `test_parser.py`: construct an `assistant.message` `SessionEvent` with `data={"outputTokens": 1234.0}` (float), call `build_session_summary` on it, and assert that `active_output_tokens == 1234` (tokens are not silently dropped).
- Add a corresponding test for `_detect_resume` with the same float token value appearing after a `session.shutdown` event.
- Verify `_safe_int_tokens` (or its replacement) still rejects `True`, `False`, `-1`, and `"500"` as invalid inputs.

> Generated by [Code Health Analysis](https://github.com/microsasa/cli-tools/actions/runs/24047741790/agentic_workflow) · ● 2.5M · [◷](https://github.com/search?q=repo%3Amicrosasa%2Fcli-tools+is%3Aissue+%22gh-aw-workflow-call-id%3A+microsasa%2Fcli-tools%2Fcode-health%22&type=issues)

### PR DESCRIPTION
**Pull request by microsasa**
Closes #810

## Problem

`_safe_int_tokens` in `parser.py` silently discards valid token counts when `outputTokens` arrives as a JSON float (e.g. `1234.0`). Python's `json` module parses `1234.0` as a `float`, but `_safe_int_tokens` only accepted `int` — causing whole-number floats to be quietly dropped with no warning.

## Fix

Replaces `_safe_int_tokens` with `_extract_output_tokens`, implementing **Option B** from the issue (preferred, guideline-aligned). The new function delegates to `AssistantMessageData.model_validate()` for type coercion:

- **Whole-number floats** (`1234.0`): Pydantic lax-mode coerces to `int` → `1234` ✅
- **Fractional floats** (`1.5`, `2.9`): Pydantic raises `ValidationError` → `None` ✅
- **Negative values** (`-1`): post-validation `> 0` check → `None` ✅
- **Null/None**: Pydantic `ValidationError` → `None` ✅

Both call sites (`_first_pass` and `_detect_resume`) are updated. The `isinstance` check is removed from business logic, aligning with the coding guideline: *"No isinstance checks in business logic."*

## Changes

- **`src/copilot_usage/parser.py`**: Replace `_safe_int_tokens` with `_extract_output_tokens`; update both call sites; add `AssistantMessageData` import
- **`tests/copilot_usage/test_parser.py`**: Replace `TestSafeIntTokens` with `TestExtractOutputTokens` (unit tests with `SessionEvent` objects); replace `TestSafeIntTokensIntegration` with `TestExtractOutputTokensIntegration` (adds regression tests for whole-number float coercion in both active and post-shutdown paths); update existing edge-case tests for Pydantic coercion behavior

## Testing

All checks pass: `make check` → lint ✅, pyright strict ✅, bandit ✅, 1135 unit tests ✅ (99% coverage), 86 e2e tests ✅

> Generated by [Issue Implementer](https://github.com/microsasa/cli-tools/actions/runs/24053203847/agentic_workflow) · ● 25.1M · [◷](https://github.com/search?q=repo%3Amicrosasa%2Fcli-tools+%22gh-aw-workflow-id%3A+issue-implementer%22&type=pullrequests)

### CODE REVIEW

File: tests/copilot_usage/test_parser.py [hunk 1]
```diff
@@ -4248,48 +4249,102 @@ def test_output_tokens_boolean_false_excluded(self, tmp_path: Path) -> None:
...
+        assert _extract_output_tokens(_make_assistant_event(42)) == 42
+
+    def test_coerces_whole_float_to_int(self) -> None:
+        """Whole-number floats like ``1234.0`` are coerced to int via Pydantic."""
+        assert _extract_output_tokens(_make_assistant_event(1234.0)) == 1234
+
+    def test_returns_none_for_fractional_float(self) -> None:
+        assert _extract_output_tokens(_make_assistant_event(3.14)) is None
 
     def test_returns_none_for_bool_true(self) -> None:
```
[the remark this answers is not in the document]
Comment (microsasa) [reply tests/copilot_usage/test_parser.py]: Fixed — the test now correctly expects `None` (booleans are rejected by the field validator), so the name `test_returns_none_for_bool_true` is accurate and the docstring is updated.

> Generated by [Review Responder](https://github.com/microsasa/cli-tools/actions/runs/24053908832/agentic_workflow) · ● 7.9M
